src/Filesystem/Filesystem.php

```diff
@@ -0,0 +1,257 @@
+<?php
+declare(strict_types=1);
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+
+namespace Cake\Filesystem;
+
+use Cake\Core\Exception\Exception;
+use CallbackFilterIterator;
+use FilesystemIterator;
+use RecursiveCallbackFilterIterator;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use RegexIterator;
+use SplFileInfo;
+use Traversable;
+
+/**
+ * @since 4.0.0
+ * @internal
+ */
+class Filesystem
+{
+    /**
+     * Direcotory type constant
+     *
+     * @var string
+     */
+    public const TYPE_DIR = 'dir';
+
+    /**
+     * Find files (non-recursively) in given directory path.
+     *
+     * @param string $path Directory path.
+     * @param mixed $filter If string will be used as regex for filtering using
+     *   `RegexIterator`, if callable will be as callback for `CallbackFilterIterator`.
+     * @param int|null $flags Flags for FilesystemIterator::__construct();
+     * @return \Traversable
+     */
+    public function find(string $path, $filter = null, ?int $flags = null): Traversable
+    {
+        $flags = $flags ?? FilesystemIterator::KEY_AS_PATHNAME
+            | FilesystemIterator::CURRENT_AS_FILEINFO
+            | FilesystemIterator::SKIP_DOTS;
+        $directory = new FilesystemIterator($path, $flags);
+
+        if ($filter === null) {
+            return $directory;
+        }
+
+        if (is_string($filter)) {
+            return new RegexIterator($directory, $filter);
+        }
+
+        return new CallbackFilterIterator($directory, $filter);
+    }
+
+    /**
+     * Find files recursively in given directory path.
+     *
+     * @param string $path Directory path.
+     * @param mixed $filter If string will be used as regex for filtering using
+     *   `RegexIterator`, if callable will be as callback for `CallbackFilterIterator`.
+     *   Hidden directories (starting with dot e.g. .git) are always skipped.
+     * @param int|null $flags Flags for FilesystemIterator::__construct();
+     * @return \Traversable
+     */
+    public function findRecursive(string $path, $filter = null, ?int $flags = null): Traversable
```

Comment:
findFilesRecursive()?

Reply:
As the signature is the same as findFIles could both be called by one proxy method having recursive = true/false as the last arg?

Reply:
Boolean args are frowned upon. A more explicit API is better.